crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+                .set_item(key, value.index().into_py(py))
+                .unwrap()
+        })
+    }
+
+    pub fn remove(&mut self, key: &PyObject) -> Option<NodeIndex> {
+        Python::with_gil(|py| -> Option<NodeIndex> {
+            let bound_dict = self.dict.bind(py);
+            let res = bound_dict
+                .get_item(key.clone_ref(py))
+                .unwrap()
+                .map(|v| NodeIndex::new(v.extract().unwrap()));
+            let _del_result = bound_dict.del_item(key);
+            res
+        })
+    }
+    pub fn values<'py>(&self, py: Python<'py>) -> impl Iterator<Item = NodeIndex> + 'py {
+        let values = self.dict.bind(py).values();
+        values.iter().map(|x| NodeIndex::new(x.extract().unwrap()))
+    }
+}
+
+/// Quantum circuit as a directed acyclic graph.
+///
+/// There are 3 types of nodes in the graph: inputs, outputs, and operations.
+/// The nodes are connected by directed edges that correspond to qubits and
+/// bits.
+#[pyclass(module = "qiskit._accelerate.circuit")]
+#[derive(Clone, Debug)]
+pub struct DAGCircuit {
+    /// Circuit name.  Generally, this corresponds to the name
+    /// of the QuantumCircuit from which the DAG was generated.
+    #[pyo3(get, set)]
+    name: Option<PyObject>,
+    /// Circuit metadata
+    #[pyo3(get, set)]
+    metadata: Option<PyObject>,
+
+    calibrations: HashMap<String, Py<PyDict>>,
+
+    pub(crate) dag: StableDiGraph<NodeType, Wire>,
+
+    #[pyo3(get)]
+    qregs: Py<PyDict>,
+    #[pyo3(get)]
+    cregs: Py<PyDict>,
+
+    /// The cache used to intern instruction qargs.
+    qargs_cache: IndexedInterner<Vec<Qubit>>,
+    /// The cache used to intern instruction cargs.
+    cargs_cache: IndexedInterner<Vec<Clbit>>,
+    /// Qubits registered in the circuit.
+    pub(crate) qubits: BitData<Qubit>,
+    /// Clbits registered in the circuit.
+    pub(crate) clbits: BitData<Clbit>,
+    /// Global phase.
+    global_phase: Param,
+    /// Duration.
+    #[pyo3(get, set)]
+    duration: Option<PyObject>,
+    /// Unit of duration.
+    #[pyo3(get, set)]
+    unit: String,
+
+    // Note: these are tracked separately from `qubits` and `clbits`
+    // because it's not yet clear if the Rust concept of a native Qubit
+    // and Clbit should correspond directly to the numerical Python
+    // index that users see in the Python API.
+    /// The index locations of bits, and their positions within
+    /// registers.
+    qubit_locations: Py<PyDict>,
+    clbit_locations: Py<PyDict>,
+
+    /// Map from qubit to input nodes of the graph.
+    qubit_input_map: IndexMap<Qubit, NodeIndex, RandomState>,
+    /// Map from qubit to output nodes of the graph.
+    qubit_output_map: IndexMap<Qubit, NodeIndex, RandomState>,
+
+    /// Map from clbit to input nodes of the graph.
+    clbit_input_map: IndexMap<Clbit, NodeIndex, RandomState>,
+    /// Map from clbit to output nodes of the graph.
+    clbit_output_map: IndexMap<Clbit, NodeIndex, RandomState>,
```

Comment:
I took a look at doing this because I liked it as a suggestion. I was looking at moving to `clbit_io_map: Vec<[NodeIndex; 2]>` . For the most part it works, but the trouble comes around removals looking at the current code there can be holes in the bit indices in the mappings when you remove idle wires from the dag. This could be addressed but would require reindexing everything in the dag on wire removal. The other option is to take a page from `StableGraph` and make the type: `Vec<Option<[NodeIndex; 2]>>` and maintain a free list which probably would be the better path. While removals are quite rare in practice and we probably don't need to worry about their overhead too much, I agree and think we should probably do this in a follow up. just to avoid adding too much complexity in this PR

Reply:
I've moved to using a vec in: https://github.com/Qiskit/qiskit/pull/12550/commits/818681a0c958177568d3b110c8881ea6ee3e7e84